x-pack/plugins/security_solution/public/management/components/endpoint_responder/hooks.tsx

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { useEffect, useRef } from 'react';
+import { useIsMounted } from '../../hooks/use_is_mounted';
+import { useGetActionDetails } from '../../hooks/endpoint/use_get_action_details';
+import { ACTION_DETAILS_REFRESH_INTERVAL } from './constants';
+import type { ActionRequestState, ActionRequestComponentProps } from './types';
+import type { useSendIsolateEndpointRequest } from '../../hooks/endpoint/use_send_isolate_endpoint_request';
+import type { useSendReleaseEndpointRequest } from '../../hooks/endpoint/use_send_release_endpoint_request';
+
+export const useUpdateActionState = ({
+  actionRequestApi,
+  actionRequest,
+  command,
+  endpointId,
+  setStatus,
+  setStore,
+  isPending,
+}: Pick<ActionRequestComponentProps, 'command' | 'setStatus' | 'setStore'> & {
+  actionRequestApi: ReturnType<
+    typeof useSendIsolateEndpointRequest | typeof useSendReleaseEndpointRequest
+  >;
+  actionRequest?: ActionRequestState;
+  endpointId?: string;
+  isPending: boolean;
+}) => {
+  const isMounted = useIsMounted();
+  const actionRequestSent = Boolean(actionRequest?.requestSent);
+  const { data: actionDetails } = useGetActionDetails(actionRequest?.actionId ?? '-', {
+    enabled: Boolean(actionRequest?.actionId) && isPending,
+    refetchInterval: isPending ? ACTION_DETAILS_REFRESH_INTERVAL : false,
+  });
+
+  // keep a reference to track the console's mounted state
+  // in order to update the store and cause a re-render on action request API response
+  const latestIsMounted = useRef(false);
```

Comment:
Why is this needed? Isn't enough with the `isMounted` ref?

Reply:
It has to do with the function inside of the `useEffect()` below and the condition we're trying to detect. Below, on line 60, we try to trigger a React re-render by updating the store for the command. Because that block of code only runs once (creates the action request) and its async, we need to ensure that when the response comes back and the `.then()` logic is triggered, that it still has a view into what the latest state of the component is. Initially, we had it using the `isMounted` property, but cause the function to `useEffect()` only has visibility into the variable that were defined at the time the function was eval()'d, `isMounted` was always set to the initial value (`false`), so the action would never update in the UI. To get around that, we ensure that the latest `isMounted` value is store instead in a mutable `ref`. so that the `then()` callback can always access it.